nssp_demo/score_forecast.R

```diff
@@ -0,0 +1,209 @@
+#'
+#' NB: this function assumes that _log-scale_ scoring is the default. If you
+#' want to vary this behaviour, you can splat additional arguments to
+#' `scoringutils::transform_forecasts` such as the identity transformation e.g.
+#' `fun = identity` with `label = "identity"`.
+#'
+#' If more than one model is present in the data, in the column `model_col` the
+#' function will add relative skill metrics to the output.
+#'
+#' @param scorable_data A data frame containing the data to be scored.
+#' @param forecast_unit A character string specifying the forecast unit.
+#' @param observed A character string specifying the column name for observed
+#' values.
+#' @param predicted A string vector specifying the column name for predicted
+#' values.
+#' @param sample_id A character string specifying the column name for sample
+#' IDs. Default is ".draw".
+#' @param model_col A character string specifying the column name for models.
+#' @param ... Additional arguments passed to
+#' `scoringutils::transform_forecasts`.
+#'
+#' @return A data frame with scored forecasts and relative skill metrics.
+#' @export
+score_single_run <- function(
+    scorable_data, quantile_only_data, forecast_unit, observed, predicted,
+    sample_id = ".draw", model_col = "model", ...) {
+  forecast_sample_df <- scorable_data |>
+    scoringutils::as_forecast_sample(
+      forecast_unit = forecast_unit,
+      observed = observed,
+      predicted = predicted,
+      sample_id = sample_id
+    )
+
+  quantile_only_df <- quantile_only_data |>
+    scoringutils::as_forecast_quantile(
+      forecast_unit = forecast_unit,
+      observed = observed,
+      predicted = predicted
+    )
+  quants <- unique(quantile_only_df$quantile_level)
+  quantiles_from_samples_df <- forecast_sample_df |>
+    scoringutils::as_forecast_quantile(probs = quants)
+
+  forecast_quantile_df <-
+    dplyr::bind_rows(
+      quantile_only_df,
+      quantiles_from_samples_df
+    ) |>
+    scoringutils::as_forecast_quantile()
+
+  sample_scores <- forecast_sample_df |>
+    scoringutils::transform_forecasts(...) |>
+    scoringutils::score()
+
+  quantile_scores <- forecast_quantile_df |>
+    scoringutils::transform_forecasts(...) |>
+    scoringutils::score()
+  # Add relative skill if more than one model is present
+  if (n_distinct(scorable_data[[model_col]]) != 1) {
+    sample_scores <- scoringutils::add_relative_skill(sample_scores)
+    quantile_scores <- scoringutils::add_relative_skill(quantile_scores)
+  }
+  return(list(
+    sample_scores = sample_scores,
+    quantile_scores = quantile_scores
+  ))
+}
+
+
+prep_truth_data <- function(truth_data_path) {
+  dat <- readr::read_tsv(truth_data_path,
+    show_col_types = FALSE
+  ) |>
+    filter(data_type == "test") |>
+    rename(true_value = ed_visits) |>
+    mutate(disease = case_when(
+      disease %in% c("Influenza", "COVID-19") ~ "Disease",
+      disease == "Total" ~ "Total",
+      TRUE ~ NA
```

Comment:
Should we warn or error if other diseases are present?

Reply:
This is done as of 927bfbd